### ISSUE TITLE
RTKQ: refetch() can't be a promise ?

### ISSUE THREAD

**Issue by vadeneev** [reduxjs/redux-toolkit#1939]
Could you help me to get why RTKQ useQuery's refetch is not a promise-like thing ? (what is motivation)
I mean - I can't await it e.t.c
same for lazyQuery and it's trigger

**Comment by phryneas** [comment]
The lazyQuery trigger can be awaited in 1.7. 
As for refetch, I'm not sure from the back of my mind - historically that was problematic in some situations when a fetch was already running, now it should be technically possible but never came up I think.

**Comment by phryneas** [comment]
@jpcody but you already have a `isFetching` property on the query hook result?

**Comment by jpcody** [comment]
Yes. which mostly fits my needs, but not always.

```tsx
// Component A - listing all dogs
const { data, isFetching, isLoading, refetch } = useGetDogsQuery()
<PullToRefresh onRefresh={refetch} refreshing={isFetching && !isLodading} />

// Component B - showing a single dog
const fetchDog, { isFetching, isLoading } = useLazyGetDogByIdQuery()
const refetch = useCallback(() => fetchDog(id), [id])
<PullToRefresh onRefresh={refetch} refreshing={isFetching && !isLodading} />
```

Component A actually behaves as expected, but component B doesn't, which leads me to want an async refetch.

In component B, the lazy query only returns `isLoading` on the first component load, not any subsequent loads of different components that use the same query. So in my case (in React Native), if I load a component that lazy queries for dog 1, then a new component that lazy queries for dog 2, the second component's `isLoading` will always return false. So on subsequent loads, `isRefreshing` will always evaluate as true and the refreshing state will appear instead of the loading state.

That lazy query behavior is unexpected to me, and perhaps I'm doing something wrong on my end. But given that behavior, in my pull-to-refresh component, I'd like to simply `await` the value I pass in `onRefresh`, which could be either a refetch function or a lazy query.

**Comment by jpcody** [comment]
Sorry, this is difficult to communicate about. What's unexpected to me is that when a component is re-mounted, and thus the query is re-issued with a different argument and different data is fetched, `isLoading` does not fire again. So after I lazily fetch 1 dog in a component in the above example, then fetching a different dog via re-mounting the same component always reports `isLoading` as false.

**Comment by awdyson** [comment]
If nothing else, consistency in RTKQ's explicit calls would be nice.
The submit method in there is ingested by a smart RHF form that disables controls while that promise is alive.
I ended up needing to use a ContextProvider to continue that disabled state through the life of the refetch.
<img width="462" alt="image" src="https://user-images.githubusercontent.com/5612248/162097294-2c70b8b0-c00e-4aca-910d-7df1253e348c.png">
<img width="327" alt="image" src="https://user-images.githubusercontent.com/5612248/162097682-588f8e58-5c33-4546-b3d1-589df6c123cf.png">
<img width="652" alt="image" src="https://user-images.githubusercontent.com/5612248/162098322-f09a55fd-62c3-4558-bb79-78da06cd098a.png">

Obviously I figured out a workaround, but hopefully this is some extra proof of value.

**Comment by phryneas** [comment]
@awdyson if nothing else, a respectful tone would be nice.

As already stated, there is no philosophical reason against doing this (it was just added in other places in later versions of RTK, but not there yet) - and thus also no reason for someone not to just whip up a PR if they need this feature. It's not a lot of code, in fact I've just done it - #2212.

Please go ahead and test the CodeSandbox build. I'm getting breakfast now.

**Comment by awdyson** [comment]
@phryneas I'm really sorry.
I spent a while trying to figure out how to concisely phrase a use-case that would help justify this ticket, and not enough time making sure I expressed how grateful for the work being done here.

In good faith, I tried my code against that PR's commit hash.
Looks like the type info is correct, but the promise is resolving immediately.
<img width="471" alt="image" src="https://user-images.githubusercontent.com/5612248/162270121-807d30a1-b2d8-4a90-a1d2-d0005bb06e2f.png">

I've got the branch pulled locally to see if I can figure out where to change that, but I often feel a little out-classed by open source libraries.
Will comment in the PR thread if I can find a solution.
